### PR TITLE
[v0.1.1]  Standardization of automation

### LINKED ISSUE #7: [Enhancement Proposal] Use loguru for better logging mechanism
Use [loguru](https://github.com/Delgan/loguru/) for better logging mechanism.

## Usage

```python
from loguru import logger

logger.info("<string>")
logger.debug("<string>")
logger.error("<string>")
```

### PR DESCRIPTION
**Pull request by NISH1001**
# Major changes

- Add `structures.TransferDTO` dataclass to store information about start/end time for each file. (This information is used in downstream analysis tasks)
- Any `AbstractAutomation`'s `run_automation(...)` method now returns `Tuple[TransferDTO]`. This will help in standardizing the execution flow
- `RcloneAutomation` now generates log and intermediate conf files to `/tmp/` prefixed as `rclone_`.
- A public interface is provided to expose any "publicly-visible" components at `api.py` (this helps to avoid polluting `__init__.py`)
- `MFTAutomation` now uses external `misc.shell.ShellExecutor` like `RcloneAutomation`.
- `source_storage_id` and `dest_storage_id` keys are to be expected in the config yaml for `MFTAutomation`.

# Minor Changes
- `controller.py` and `controller_test.py` has been removed from the main package.
- All the log parsing SLOCs has been moved into `parse_log(...)` methods in the corresponding automation component
- There's a naive automation pipeline at `tests/controller_test.py` where we iterate through all the automation components and compute corresponding throughput. 
- `tests/controller_test.py` gets all the installation directories for nifi/mft from env variables.

# TODO
- Create a separate controller component to add any number of automation and perform downstream analyses. 
- Better handling of logs (Issue at https://github.com/NASA-IMPACT/data-transfer-evaluation/issues/10)
- Selective transfer (at https://github.com/NASA-IMPACT/data-transfer-evaluation/issues/11)

@xhagrg @freitagb  @smarru Can you please review this and let me know.

### CODE REVIEW

File: evalit/mft/mft_automation.py [hunk 1]
```diff
@@ -102,25 +111,19 @@ def run_automation(self, njobs: int = 4):
...
                 for part in exdto.output:
                     if part.find("STARTING") != -1:
-                        start_time = part.split("|")[1].strip()
-                        start_time_map[transfer_id] = int(start_time) / 1000
-                        logger.debug("Start time: ", start_time)
+                        dto = timekeeper.get(
+                            transfer_id,
+                            TransferDTO(fname=transfer_id, transferer="mft"),
+                        )
+                        dto.start_time = int(part.split("|")[1].strip()) / 1000
```
Comment (code-geek) [inline evalit/mft/mft_automation.py:118]: A quick comment on what's expected in `part` would be helpful.

File: evalit/nifi/nifi_automation.py [hunk 2]
```diff
@@ -33,9 +35,13 @@ def __init__(
     ) -> None:
         super().__init__(config=config, files=files, debug=debug)
         self.nifi_url = nifi_url
+
+        assert os.path.exists(nifi_dir), f"{nifi_dir} path doesn't exist!"
```
Comment (code-geek) [inline evalit/nifi/nifi_automation.py:39]: Just curious: How about a `try/except` instead?

File: evalit/nifi/nifi_automation.py [hunk 3]
```diff
@@ -364,44 +370,43 @@ def run_automation(self):
...
-        while 1:
-            # time.sleep(5)
+        # TODO: Optimize this!
+        while (len(timekeeper) < nfiles) and (end_counter < nfiles):
+            time.sleep(poll_wait_time)
             # read file line by line
-            with open(log_file_location, "r") as f:
+            with open(log, "r") as f:
                 for line in f:
                     if line.find(session_uuid) != -1:
```
Comment (code-geek) [inline evalit/nifi/nifi_automation.py:396]: Could we use `.readlines()`?

File: evalit/nifi/nifi_automation.py [hunk 4]
```diff
@@ -364,44 +370,43 @@ def run_automation(self):
...
-                        if startLine:
-                            start_time_map[file_name] = utc_time
-                        else:
-                            end_time_map[file_name] = utc_time
-
-                print(start_time_map)
-                if len(end_time_map) == total_files:
-                    break
+                        fname = line.split(session_uuid)[1].strip()
+                        is_start = line.split(session_uuid)[0].find("Starting") > -1
```
Comment (code-geek) [inline evalit/nifi/nifi_automation.py:401]: An example of `line` would make this code easier to understand

File: evalit/rclone/rclone_automation.py [hunk 5]
```diff
@@ -107,56 +106,62 @@ def run_automation(self):
...
+                fname = line.split(":")[3].strip()
+                dto = timekeeper.get(
+                    fname, TransferDTO(fname=fname, transferer="rclone")
+                )
+                dto.start_time = datetime.strptime(
+                    line.split("DEBUG")[0].strip(), "%Y/%m/%d %H:%M:%S"
+                )
+                timekeeper[fname] = dto
+
+            if line.find("Copied") != -1:
```
Comment (code-geek) [inline evalit/rclone/rclone_automation.py:155]: Should we use a constant for -1? Like `NOT_FOUND_ERROR_CODE = -1`

`.find()` is common enough that people should know this automatically, but the -1 does make a casual reader pause for a second.

File: evalit/rclone/rclone_automation.py [hunk 6]
```diff
@@ -107,56 +106,62 @@ def run_automation(self):
...
+                fname = line.split(":")[3].strip()
```
Comment (code-geek) [inline evalit/rclone/rclone_automation.py:156]: Why 3?

File: tests/controller_test.py [hunk 7]
```diff
@@ -8,65 +9,79 @@
...
+    )
+    times = dtotimes_to_times(timesdto)
     total_valume = 0.0
     for i in range(len(file_sizes)):
         total_valume += file_sizes[i]
-    throughput = total_valume * 8 / (np.max(times) - np.min(times))
-    print("Throughput: " + str(throughput))
+    val = 0
+    try:
+        val = total_valume * 8 / (np.max(times) - np.min(times))
```
Comment (code-geek) [inline tests/controller_test.py]: nitpick: valume -> volume?

File: tests/controller_test.py [hunk 8]
```diff
@@ -8,65 +9,79 @@
...
 
 # TODO: Fetch theses values from the S3 python client
 file_sizes = [1]  # GB
 file_list = ["testfile"]  # File list in source bucket
 
 # nifi_installation = "/sproj/MFT/nifi-1.15.3"
-nifi_installation = "/home/nishan/software/nifi/nifi-1.15.3"
-mft_installation = "/proj/MFT/build"
+nifi_installation = os.getenv(
+    "NIFI_INSTALLATION", "/home/nishan/software/nifi/nifi-1.15.3"
```
Comment (code-geek) [inline tests/controller_test.py]: Can we avoid this hardcoding?

File: evalit/nifi/nifi_automation.py (hunk 2 above)
Comment (NISH1001) [reply evalit/nifi/nifi_automation.py:39]: This is meant as a sanity check that nifi_dir should pre-exist. So I think try/catch doesn't make sense.

File: evalit/nifi/nifi_automation.py (hunk 3 above)
Comment (NISH1001) [reply evalit/nifi/nifi_automation.py:396]: Oh ya. I just changed the log file name here. Will do that. (Previous code)

File: evalit/nifi/nifi_automation.py (hunk 4 above)
Comment (NISH1001) [reply evalit/nifi/nifi_automation.py:401]: Ah ya. I'd check the log and see. Thanks.

File: evalit/rclone/rclone_automation.py (hunk 6 above)
Comment (NISH1001) [reply evalit/rclone/rclone_automation.py:156]: Specific to rclone log. Will have to check thoroughly if we can make it flexible. Probably, filename exists at that index for the line.

File: tests/controller_test.py (hunk 8 above)
Comment (NISH1001) [reply tests/controller_test.py]: Thanks. Will do. I wanted to add fallback value for my device as a test.

File: evalit/nifi/nifi_automation.py (hunk 3 above)
Comment (NISH1001) [reply evalit/nifi/nifi_automation.py:396]: I think `.readlines()` is an eager one. Returns a list. So, normally I prefer iterating through the file pointer.

**Pull request merged**